File: library/src/main/java/com/ebr163/inspector/value/InspectVariableBuilder.java
package com.ebr163.inspector.value;

import com.ebr163.inspector.rule.Rule;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by Bakht
 * on 09.01.2018.
 */

public class InspectVariableBuilder<Type> {

    private Type value;
    private List<Rule<Type>> rules;
    private InspectionVariable.OnErrorListener errorListener;

    public InspectVariableBuilder(Type value) {
        this.value = value;
        rules = new ArrayList<>();
    }

    public InspectVariableBuilder<Type> addRule(Rule<Type> rule) {
        rules.add(rule);
        return this;
    }

    @SafeVarargs
    public final InspectVariableBuilder<Type> addRules(Rule<Type>... rules) {
        Collections.addAll(this.rules, rules);
        return this;
    }

    public InspectVariableBuilder<Type> addErrorListener(InspectionVariable.OnErrorListener errorListener) {
        this.errorListener = errorListener;
        return this;
    }

    public InspectionVariable<Type> build() {
        return new InspectionVariable<>(value, rules, errorListener);
    }
}

File: library/src/main/java/com/ebr163/inspector/variable/InspectVariableBuilder.kt
package com.ebr163.inspector.variable


import com.ebr163.inspector.rule.Rule
import java.util.Collections

/**
 * Created by Bakht
 * on 09.01.2018.
 */

class InspectVariableBuilder<Type>(private val value: Type?) {

    private val rules: MutableList<Rule<Type>> = mutableListOf()
    private var errorListener: ((enabled: Boolean, errorMessage: String?) -> Unit)? = null

    fun addRule(rule: Rule<Type>): InspectVariableBuilder<Type> {
        rules.add(rule)
        return this
    }

    @SafeVarargs
    fun addRules(vararg rules: Rule<Type>): InspectVariableBuilder<Type> {
        Collections.addAll(this.rules, *rules)
        return this
    }

    fun addErrorListener(errorListener: (Boolean, String?) -> Unit): InspectVariableBuilder<Type> {
        this.errorListener = errorListener
        return this
    }

    fun build(): InspectionVariable<Type> = InspectionVariable(value, rules, errorListener)
}